posts/zzz_DO_NOT_EDIT_the__tensio.../the__tension_of__high-_performance__computing:__reproducibility_vs.__parallelization.qmd

```diff
@@ -0,0 +1,259 @@
+reproducibility, especially with random number generation (RNG) or tasks needing
+strict order.
+
+To enhance reproducibility, `{mirai}` allows disabling the dispatcher, directly
+connecting the host to daemons in a round-robin fashion. While less efficient, 
+this approach provides greater control over task execution and is better suited
+for ensuring reproducibility by initializing 
+[L'Ecuyer-CMRG RNG streams](https://pubsonline.informs.org/doi/10.1287/opre.47.1.159).
+
+```{r}
+library(mirai)
+library(dplyr, warn.conflicts = FALSE)
+
+# Parameters for the simulation
+cohorts <- tribble(
+  ~patient_count, ~mean_effect, ~sd_effect,
+  1000,           0.7,          0.1,
+  1000,           0.65,         0.15,
+  1000,           0.75,         0.05
+)
+
+# Start daemons with consistent RNG streams
+x <- daemons(4, dispatcher = FALSE, seed = 123)
+
+# Parallel simulation for each row of the cohorts table
+m <- mirai_map(cohorts, \(patient_count, mean_effect, sd_effect) {
+  dplyr::tibble(
+    patient_id = 1:patient_count,
+    efficacy = rnorm(patient_count, mean = mean_effect, sd = sd_effect)
+  )
+})
+
+results <- m[] |> dplyr::bind_rows()
+
+x <- daemons(0, dispatcher = FALSE)
+
+results %>%
+  group_by(patient_id) %>%
+  summarise(
+    mean_efficacy = mean(efficacy),
+    sd_efficacy = sd(efficacy)
+  )
+```
+
+In this example, we use `tribble()` to define the simulation parameters and 
+initialize 4 daemons with `dispatcher = FALSE` and a fixed seed to ensure 
+consistent random number generation across tasks. The `mirai_map()` function
+parallelizes the drug efficacy simulation, and the results are combined using 
+`bind_rows()` for further analysis. Disabling the dispatcher gives more control
+over task execution, ensuring reproducibility. If you repeat the computation you
+will notice that it generates consistent results.
+
+However, this approach comes at a cost. Disabling the dispatcher may lead to
+inefficient resource utilization when tasks are unevenly distributed, as some 
+daemons may remain idle. While reproducibility is prioritized, we sacrifice
+some performance, especially when handling tasks with varying workloads.
+
+> "In prioritizing reproducibility, we inevitably sacrifice some performance, 
+especially when tasks with unequal workloads are distributed across daemons."
+
+Reproducibility becomes more complex when using parallelization frameworks
+like `{parallelMap}`, `{doFuture}`, and `{future}`, as each handles random 
+number generation (RNG) differently. While `set.seed()` is sufficient for 
+sequential tasks, parallel computations require managing RNG streams carefully, 
+often using types like "L'Ecuyer-CMRG" or functions such as 
+`clusterSetRNGStream()` for synchronization. Each framework requires 
+specific adjustments to ensure consistent results, emphasizing the
+importance of understanding how each backend manages RNG in parallel
+environments.
+
```

Comment:
I would like to add to this that even without random numbers and for seemingly simple tasks such as summing over floating point numbers the results may differ because floating point numbers are not represented exactly. The exact sum depends on the order of operations.

Reply:
Great point! Added a small paragraph after this.